langchain/src/vectorstores/analyticdb.ts

```diff
@@ -0,0 +1,290 @@
+      throw new Error(
+        `Vectors must have the same length as the number of dimensions (${this.embeddingDimension})`
+      );
+    }
+
+    if (!this.isCreateCollection) {
+      await this.createCollection();
+    }
+
+    const client = await this.pool.connect();
+    try {
+      const chunkSize = 500;
+      const chunksTableData: DataType[] = [];
+
+      for (let i = 0; i < documents.length; i += 1) {
+        chunksTableData.push({
+          id: uuid.v4(),
+          embedding: vectors[i],
+          document: documents[i].pageContent,
+          metadata: documents[i].metadata,
+        });
+
+        // Execute the batch insert when the batch size is reached
+        if (chunksTableData.length === chunkSize) {
+          const rs = new Readable();
+          let currentIndex = 0;
+          rs._read = function () {
+            if (currentIndex === chunkSize) {
+              rs.push(null);
+            } else {
+              const data = chunksTableData[currentIndex];
+              rs.push(
+                `${data.id}\t{${data.embedding.join(",")}}\t${
+                  data.document
+                }\t${JSON.stringify(data.metadata)}\n`
+              );
+              currentIndex += 1;
+            }
+          };
+          const ws = client.query(
+            copyFrom(
+              `COPY ${this.collectionName}(id, embedding, document, metadata) FROM STDIN`
+            )
+          );
+
+          await pipeline(rs, ws);
+          // Clear the chunksTableData list for the next batch
+          chunksTableData.length = 0;
+        }
+      }
+
+      // Insert any remaining records that didn't make up a full batch
+      if (chunksTableData.length > 0) {
+        const rs = new Readable();
+        let currentIndex = 0;
+        rs._read = function () {
+          if (currentIndex === chunksTableData.length) {
+            rs.push(null);
+          } else {
+            const data = chunksTableData[currentIndex];
+            rs.push(
+              `${data.id}\t{${data.embedding.join(",")}}\t${
+                data.document
+              }\t${JSON.stringify(data.metadata)}\n`
+            );
+            currentIndex += 1;
+          }
+        };
+        const ws = client.query(
+          copyFrom(
+            `COPY ${this.collectionName}(id, embedding, document, metadata) FROM STDIN`
+          )
+        );
+        await pipeline(rs, ws);
+      }
+    } finally {
+      client.release();
+    }
+  }
+
+  async similaritySearchVectorWithScore(
+    query: number[],
+    k: number,
+    filter?: this["FilterType"]
+  ): Promise<[Document, number][]> {
+    if (!this.isCreateCollection) {
+      await this.createCollection();
+    }
+
+    let filterCondition = "";
+    const filterEntries = filter ? Object.entries(filter) : [];
+    if (filterEntries.length > 0) {
+      const conditions = filterEntries.map(
+        ([key, value]) => `metadata->>'${key}' = '${value}'`
```

Comment:
Hey actually I think this is a SQL injection risk?

Reply:
I think something like this is better: https://github.com/hwchase17/langchainjs/blob/main/langchain/src/vectorstores/typeorm.ts#L137

Reply:
Thanks for correcting!